### PR TITLE
[WIP][master] pylint check

### PR DESCRIPTION
**Pull request by fmessmer**
still WIP, but I wanted to hear people's thoughts about adding `pylint` to `industrial_ci`

example config, i.e. add to `.travis.yml` of your repo which contains python code
```
env:
  global:
    - PYLINT=true
    - PYLINT_ARGS="--output-format=parseable --reports=no"
    - PYLINT_VERSIONS="pylint pylint3"
```

motivation:
we want to verify that our code is both python2- and python3-compatible!
which is why we need to run both `pylint` (for python2.7) and `pylint3` (for python3.5)

what do you think of this?
any comments on how it should be implemented or how the this first draft should be modified?

### CODE REVIEW

File: industrial_ci/src/tests/source_tests.sh [hunk 1]
```diff
@@ -139,11 +139,22 @@ function run_source_tests {
...
     if [ "${CLANG_TIDY:-false}" != false ]; then
         run_clang_tidy_check "$target_ws"
     fi
+    if [ "${PYLINT:-false}" != false ]; then
+        local -a pylint_versions
+        ici_parse_env_array pylint_versions PYLINT_VERSIONS
+        local -a pylint_args
+        ici_parse_env_array pylint_args PYLINT_ARGS
+
+        ici_run "install_pylint" ici_install_pkgs_for_command pylint "${pylint_versions[@]}"
```
Comment (ipa-mdl) [inline industrial_ci/src/tests/source_tests.sh]: This must be run for every element in `pylint_versions`, it should be moved into the for loop.

File: industrial_ci/src/tests/source_tests.sh [hunk 2]
```diff
@@ -139,11 +139,22 @@ function run_source_tests {
...
+
+        for cmd in "${pylint_versions[@]}"; do
+            ici_run "pylint" "$cmd" "${pylint_args[@]}" "$(find "$target_ws/src" -iname "*.py")"
```
Comment (ipa-mdl) [inline industrial_ci/src/tests/source_tests.sh]:
```suggestion
            ici_run "install_$cmd" ici_install_pkgs_for_command "$cmd" "$cmd"
            ici_run "run_$cmd" "$cmd" "${pylint_args[@]}" "$(find "$target_ws/src" -iname "*.py")"
```

**Comment by fmessmer** [comment]
I started the standalone version in a new branch - see https://github.com/ros-industrial/industrial_ci/pull/464

**Comment by fmessmer** [comment]
it's still not working - will continue tomorrow...
 - with https://github.com/ros-industrial/industrial_ci/pull/463/commits/1d930cfa93c9b3d85285ae8cb21f28718f6cd7c9, I get:
```
Starting function 'run_pylint'
No config file found, using default configuration
************* Module sample_talker
E:  3, 0: Unable to import 'std_msgs.msg' (import-error)
```
 - with https://github.com/ros-industrial/industrial_ci/pull/463/commits/73fc80adb0fed41c6b3a3da7f5747fa3393e8d80, I get:
```
/root/ici/util.sh: line 80: !name: unbound variable
/root/ici/util.sh: line 181: ICI_FOLD_NAME: unbound variable
Tests failed exit code '1'
```


I have a few more things, that I would like to see discussed....will document tomorrow...

File: industrial_ci/src/tests/source_tests.sh [hunk 3]
```diff
@@ -139,11 +139,21 @@ function run_source_tests {
...
     fi
+    if [ "${PYLINT:-false}" != false ]; then
+        local -a pylint_versions
+        ici_parse_env_array pylint_versions PYLINT_VERSIONS
+        local -a pylint_args
+        ici_parse_env_array pylint_args PYLINT_ARGS
+
+        for cmd in "${pylint_versions[@]}"; do
+            ici_run "install_$cmd" ici_install_pkgs_for_command "$cmd" "$cmd"
+            ici_source_setup "$target_ws/install" ici_run "run_$cmd" "$cmd" "${pylint_args[@]}" "$(find "$target_ws/src" -iname "*.py")"
```
Comment (ipa-mdl) [inline industrial_ci/src/tests/source_tests.sh]:
```suggestion
            ici_run "run_$cmd" ici_exec_in_workspace "$extend" "$target_ws" "$cmd" "${pylint_args[@]}" "$(find "$target_ws/src" -iname "*.py")"
"${catkin_lint_args[@]}" src
            
```

**Comment by ipa-mdl** [comment]
Please reduce the Travis jobs to just your new test jobs.
Otherwise all `ros-industrial` gets blocked 1h for each push.

**Comment by gavanderhoorn** [comment]
General question / something to consider: should `pylint` be integrated with `industrial_ci`, or would it make more sense to use an approach similar to [ament_lint](https://github.com/ament/ament_lint) which is something other pkgs would depend on and the tests would be part of the regular test run instead of being 'treated special' by the CI?

Like this in [ros2/rclcpp/rclcpp_lifecycle/CMakeLists.txt](https://github.com/ros2/rclcpp/blob/b25213a186add0ee93e270d08c9cb5601ae88066/rclcpp_lifecycle/CMakeLists.txt#L46-L48).

File: .travis.yml [hunk 4]
```diff
@@ -17,67 +17,73 @@ env:
...
 
-    # Tidy
-    - ROS_DISTRO=melodic TARGET_WORKSPACE='industrial_ci/mockups/test_clang_tidy' CLANG_TIDY=pedantic
-    - ROS_DISTRO=melodic TARGET_WORKSPACE='industrial_ci/mockups/test_clang_tidy' CLANG_TIDY=pedantic CLANG_TIDY_ARGS="-checks=modernize-use-nullptr" EXPECT_EXIT_CODE=1
+    ## Tidy
+    #- ROS_DISTRO=melodic TARGET_WORKSPACE='industrial_ci/mockups/test_clang_tidy' CLANG_TIDY=pedantic
+    #- ROS_DISTRO=melodic TARGET_WORKSPACE='industrial_ci/mockups/test_clang_tidy' CLANG_TIDY=pedantic CLANG_TIDY_ARGS="-checks=modernize-use-nullptr" EXPECT_EXIT_CODE=1
+
+    # Pylint
+    - ROS_DISTRO=kinetic TARGET_WORKSPACE="industrial_ci/mockups/test_pylint" PYLINT=true PYLINT_VERSIONS="pylint pylint3"
```
Comment (fmessmer) [inline .travis.yml]: seems like
```
Starting function 'run_pylint'
No config file found, using default configuration
************* Module sample_talker
E:  3, 0: Unable to import 'std_msgs.msg' (import-error)
```
was mainly due to the `TARGET_WORKSPACE` not being set up correctly...

File: industrial_ci/src/tests/source_tests.sh [hunk 5]
```diff
@@ -97,6 +97,29 @@ function run_clang_tidy_check {
...
+    local -a pylint_versions
+    ici_parse_env_array pylint_versions PYLINT_VERSIONS
+    local -a pylint_args
+    ici_parse_env_array pylint_args PYLINT_ARGS
+
+    for cmd in "${pylint_versions[@]}"; do
+        ici_run "install_$cmd" ici_install_pkgs_for_command "$cmd" "$cmd"
+        set +e
+        ici_with_ws "$target_ws" ici_run "run_$cmd" ici_exec_in_workspace "$extend" "$target_ws" "$cmd" "${pylint_args[@]}" "$(find "$target_ws/src" -iname "*.py")"; status=$?
+        set -e
```
Comment (fmessmer) [inline industrial_ci/src/tests/source_tests.sh]: this is not yet doing the right thing
I wanted to save the exit code of each `pylint_version` to `status` and only exit with exit_code `1` at the end if one of them returned with exit code !=0

how can I get can I access the exit code of `cmd`?

Comment (ipa-mdl) [reply industrial_ci/src/tests/source_tests.sh]: Try something like https://github.com/ros-industrial/industrial_ci/blob/master/industrial_ci/src/tests/clang_format_check.sh#L36

File: .travis.yml [hunk 6]
```diff
@@ -17,67 +17,73 @@ env:
...
 
-    # Tidy
-    - ROS_DISTRO=melodic TARGET_WORKSPACE='industrial_ci/mockups/test_clang_tidy' CLANG_TIDY=pedantic
-    - ROS_DISTRO=melodic TARGET_WORKSPACE='industrial_ci/mockups/test_clang_tidy' CLANG_TIDY=pedantic CLANG_TIDY_ARGS="-checks=modernize-use-nullptr" EXPECT_EXIT_CODE=1
+    ## Tidy
+    #- ROS_DISTRO=melodic TARGET_WORKSPACE='industrial_ci/mockups/test_clang_tidy' CLANG_TIDY=pedantic
+    #- ROS_DISTRO=melodic TARGET_WORKSPACE='industrial_ci/mockups/test_clang_tidy' CLANG_TIDY=pedantic CLANG_TIDY_ARGS="-checks=modernize-use-nullptr" EXPECT_EXIT_CODE=1
+
+    # Pylint
+    - ROS_DISTRO=kinetic TARGET_WORKSPACE="industrial_ci/mockups/test_pylint" PYLINT=true PYLINT_VERSIONS="pylint pylint3" PYLINT_ARGS="--reports=no"
```
Comment (ipa-mdl) [inline .travis.yml]: If you only add one entry, Travis does not build the job matrix for some reason

File: industrial_ci/src/tests/source_tests.sh [hunk 7]
```diff
@@ -97,6 +97,33 @@ function run_clang_tidy_check {
...
+    local status=0
+    local exit_code=0
+
+    local -a pylint_versions
+    ici_parse_env_array pylint_versions PYLINT_VERSIONS
+    local -a pylint_args
+    ici_parse_env_array pylint_args PYLINT_ARGS
+
+    for cmd in "${pylint_versions[@]}"; do
+        ici_run "install_$cmd" ici_install_pkgs_for_command "$cmd" "$cmd"
```
Comment (ipa-mdl) [inline industrial_ci/src/tests/source_tests.sh]: Do not show the install process by default:
```suggestion
        ici_quiet ici_install_pkgs_for_command "$cmd" "$cmd"
```
Or split these loops.

**Comment by fmessmer** [comment]
@ipa-mdl 
for the time being, I'm happy with this version now....it covers our current usecase....
I will have to test it with more complex builds and report back

feel free to drop any review/remark you might have...

**Comment by fmessmer** [comment]
something is wrong with the globbing....trying to fix tomorrow:
```
Starting function 'run_pylint3'

pylint_args: --output-format=parseable

find: unknown predicate `--errors-only/src'

/root/ici/workspace.sh: line 231: cd: --: invalid option

cd: usage: cd [-L|[-P [-e]] [-@]] [dir]

pylint3 check failed with status 2

<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<

Function 'run_pylint3' returned with code '2' after 0 min 0 sec

pylint check(s) failed.

The command ".industrial_ci/travis.sh" exited with 1.
```

**Comment by fmessmer** [comment]
I'm not yet happy with the output of the `run_pylint` step...I don't see what went wrong in a more complex build...it also differs from what I see when running the `pylint` command locally in my workspace directly....investigating...

**Comment by fmessmer** [comment]
@ipa-mdl 
is it correct, that `industrial_ci` gets cloned into `/root/target_ws/src/MY_REPO/.industrial_ci`?
because with the current `(find "$target_ws/src" -type f -iname "*.py")` while-read-loop, also the python file in the `industrial_ci/mockups` are found and thus `pylint`-checked...see `pylint3 check for '/root/target_ws/src/cob_control/.industrial_ci/industrial_ci/mockups/testpkg_broken_install/script/sample_talker.py' passed`

**Comment by ipa-mdl** [comment]
> is it correct, that industrial_ci gets cloned/copied into
/root/target_ws/src/MY_REPO/.industrial_ci?

Yes, if you clone it into `.industrial_ci` in the Travis yml ;)

> this does not happen in case the travis job explicitly sets TARGET_WORKSPACE - either via *.rosinstall

Yes, because the target repo will not be copied in this case..

> because with the current (find "$target_ws/src" -type f -iname "*.py") while-read-loop, also the python file in the industrial_ci/mockups are found and thus pylint-checked

Example: https://github.com/ros-industrial/industrial_ci/blob/master/industrial_ci/src/tests/clang_format_check.sh#L36

**Comment by ipa-mdl** [comment]
@fmessmer: I implemented `ici_find` in #468. Please have a look

**Comment by fmessmer** [comment]
There are a [couple of ipa320 repos](https://github.com/pulls?utf8=%E2%9C%93&q=is%3Aopen+is%3Apr+archived%3Afalse+%5Bci_updates%5D+pylint+%2B+Python3+compatibility+user%3Aipa320+) using this feature branch where it can be seen what kind of output is produced...

It'd be great if you could have a look whether the reports is what would be expected...it slightly differs from what is produced by running `pylint` in the local source space and/or using the VisualStudioCode plugin...
I'm not sure whether all reported errors are correct positives or whether additional settings/options need to be configured to make the reports valid for a ROS/catkin workspace...

what would it take to finalize this PR in order to get it merged?

Thanks!

**Comment by fmessmer** [comment]
@ipa-mdl 
I rebased this PR onto the latest updates...curiously waiting for your review... :wink:

**Comment by ipa-mdl** [comment]
As discussed: https://github.com/ros-industrial/industrial_ci/issues/342#issuecomment-456228021

**Comment by fmessmer** [comment]
we wont be merging this PR...but rather use #477 instead....
but as I'm using `fmessmer:master_pylint` branch in several repos (ipa320, mojin-robotics), I'll have to push an updated version to this branch - overwriting this PR
I stored this version on my branch `fmessmer:master_pylint_pr463`

**Pull request closed without merging**